src/tree/RefreshManager.ts

```diff
@@ -0,0 +1,275 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { AzExtTreeItem, callWithTelemetryAndErrorHandling, IActionContext, parseError, registerCommand } from '@microsoft/vscode-azext-utils';
+import * as os from 'os';
+import * as vscode from 'vscode';
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import { EventAction, EventType, isCancellationError } from '../runtimes/docker';
+import { debounce } from '../utils/debounce';
+import { AllTreePrefixes, TreePrefix } from './TreePrefix';
+
+const pollingIntervalMs = 60 * 1000; // One minute
+const eventListenerTries = 3; // The event listener will try at most 3 times to connect for events
+const debounceDelayMs = 500; // Refreshes rapidly initiated for the same tree view will be debounced to occur 500ms after the last initiation
+
+type RefreshTarget = AzExtTreeItem | TreePrefix;
+type RefreshReason = 'interval' | 'event' | 'config' | 'manual' | 'contextChange';
+
+const ContainerEventActions: EventAction[] = ['create', 'destroy', 'die', 'kill', 'pause', 'rename', 'restart', 'start', 'stop', 'unpause', 'update'];
+const ImageEventActions: EventAction[] = ['delete', 'import', 'load', 'pull', 'save', 'tag', 'untag'];
+const NetworkEventActions: EventAction[] = ['create', 'destroy', 'remove'];
+const VolumeEventActions: EventAction[] = ['create', 'destroy'];
+
+export class RefreshManager extends vscode.Disposable {
+    private readonly disposables: vscode.Disposable[] = [];
+    private readonly cts: vscode.CancellationTokenSource = new vscode.CancellationTokenSource();
+
+    public constructor() {
+        super(() => vscode.Disposable.from(...this.disposables).dispose());
+
+        // VSCode does *not* cancel by default on disposal of a CancellationTokenSource, so we need to manually cancel
+        this.disposables.unshift(new vscode.Disposable(() => this.cts.cancel()));
+
+        this.setupRefreshOnInterval();

```

Comment:
An ill effect of this is that we're now initiating communication with Docker before needed, which would regress #3475. Need to fix this, so that none of these events are set up until the tree is refreshed for the first time.